kernel/thor/arch/x86/vmx.cpp

```
@@ -350,7 +350,36 @@ namespace thor::vmx {
 				frigg::infoLogger() << "vmx: hlt" << frigg::endLog;
 				exitInfo.exitReason = khelVmexitHlt;
 				return exitInfo;
-			} else if(reason == VMEXIT_EXTERNAL_INTERRUPT) {
+			} else if(reason == 48) {
+				frigg::UnsafePtr<Thread> this_thread = getCurrentThread();
+				size_t address = vmread(0x00002400);
+				frigg::infoLogger() << "vmx: ept violation at address: " << address << frigg::endLog;
+				WorkScope wqs{this_thread->pagingWorkQueue()};
+				struct Closure {
+					ThreadBlocker blocker;
+					Worklet worklet;
+					FaultNode fault;
+				} closure;
+
+				closure.worklet.setup([] (Worklet *base) {
+					auto closure = frg::container_of(base, &Closure::worklet);
+					Thread::unblockOther(&closure->blocker);
+				});
+				closure.fault.setup(&closure.worklet);
+				closure.blocker.setup();
+				size_t exitFlags = vmread(0x00006400);
+				uint32_t flags = 0;
+				if(exitFlags & 1) {
+					flags |= AddressSpace::kFaultWrite;
+				}
+				if(exitFlags & (1 << 2)) {
+					flags |= AddressSpace::kFaultExecute;
+				}
+				if(!space->handleFault(address, flags, &closure.fault))
+					Thread::blockCurrent(&closure.blocker);
+
+				bool handled = closure.fault.resolved();
```

What if `!handled`?

---

I guess we should return an error to userspace.